lib/prototypes/collections.js

```diff
@@ -0,0 +1,100 @@
+'use strict';
+
+const { isType } = require('../util.js');
+
+const object = {
+  rules: ['length'],
+  type: 'object',
+  kind: 'struct',
+  checkType(source, path, root) {
+    if (!this.isInstance(source)) {
+      return `Filed "${path}" is not a ${this.type}`;
+    }
+    const entries = this.entries(source);
+    if (entries.length === 0 && this.required) {
+      return `Filed "${path}" is required`;
+    }
+    for (const [field, val] of entries) {
+      if (typeof field !== this.key) {
+        return `In ${this.type} "${path}": type of key must be a ${this.key}`;
+      }
+      const error = this.value.check(val, path, root);
+      if (error.length > 0) return error;
+    }
+  },
+  construct(def, prep) {
+    const { type } = this;
+    const { [type]: short, key, value } = def;
+    const [[k, v]] = short ? Object.entries(short) : [[key, value]];
+    this.key = k;
+    const child = prep.parse(v);
+    if (isType(child)) {
+      this.value = child;
+      this.metadataFrom(child);
+    } else {
+      const schema = new prep.types.schema(child.defs, prep);
+      this.references.add('schema');
+      this.metadataFrom(schema);
+      this.value = schema;
+    }
+  },
+  isInstance(value) {
+    return typeof value === 'object';
+  },
+  entries(value) {
+    return Object.entries(value);
+  },
+};
+
+const map = {
+  ...object,
+  type: 'map',
+  isInstance(value) {
+    return Reflect.getPrototypeOf(value) === Map.prototype;
```

Comment:
This would probably not work in metavm contexts and same goes for other such checks, since they prototypes from different sandboes will probably not be reference-equal. Have you verifies this case? (`schema` created in one sandbox and `value` created in another)

Reply:
This is from old version. Replaced with checking constructor name

Reply:
Done